File: slang-checks/src/main/java/org/sonarsource/slang/checks/IfConditionalAlwaysTrueOrFalseCheck.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.checks;

import org.sonarsource.slang.api.BinaryExpressionTree.Operator;
import org.sonarsource.slang.api.IdentifierTree;
import org.sonarsource.slang.api.IfTree;
import org.sonarsource.slang.api.LiteralTree;
import org.sonarsource.slang.api.Tree;
import org.sonarsource.slang.checks.api.InitContext;
import org.sonarsource.slang.checks.api.SlangCheck;
import org.sonarsource.slang.checks.utils.ExpressionUtils;
import java.util.function.Predicate;
import org.sonar.check.Rule;

import static org.sonarsource.slang.api.BinaryExpressionTree.Operator.CONDITIONAL_AND;
import static org.sonarsource.slang.api.BinaryExpressionTree.Operator.CONDITIONAL_OR;
import static org.sonarsource.slang.checks.utils.ExpressionUtils.isBinaryOperation;
import static org.sonarsource.slang.checks.utils.ExpressionUtils.isBooleanLiteral;
import static org.sonarsource.slang.checks.utils.ExpressionUtils.isFalseValueLiteral;
import static org.sonarsource.slang.checks.utils.ExpressionUtils.isNegation;
import static org.sonarsource.slang.checks.utils.ExpressionUtils.isTrueValueLiteral;
import static org.sonarsource.slang.checks.utils.ExpressionUtils.skipParentheses;

@Rule(key = "S1145")
public class IfConditionalAlwaysTrueOrFalseCheck implements SlangCheck {

  public static final String MESSAGE_TEMPLATE = "Remove this useless \"%s\" statement.";

  @Override
  public void initialize(InitContext init) {
    init.register(IfTree.class, (ctx, ifTree) -> {
      Tree condition = ifTree.condition();
      if (isAlwaysTrueOrFalse(condition)) {
        String message = String.format(MESSAGE_TEMPLATE, ifTree.ifKeyword().text());
        ctx.reportIssue(condition, message);
      }
    });
  }

  private static boolean isAlwaysTrueOrFalse(Tree originalCondition) {
    Tree condition = skipParentheses(originalCondition);
    return isBooleanLiteral(condition)
      || isTrueValueLiteral(condition)
      || isFalseValueLiteral(condition)
      || isSimpleExpressionWithLiteral(condition, CONDITIONAL_AND, ExpressionUtils::isFalseValueLiteral)
      || isSimpleExpressionWithLiteral(condition, CONDITIONAL_OR, ExpressionUtils::isTrueValueLiteral);
  }

  private static boolean isSimpleExpressionWithLiteral(Tree condition, Operator operator, Predicate<? super Tree> hasLiteralValue) {
    boolean simpleExpression = isBinaryOperation(condition, operator)
      && condition.descendants()
        .map(ExpressionUtils::skipParentheses)
        .allMatch(tree -> tree instanceof IdentifierTree
          || tree instanceof LiteralTree
          || isNegation(tree)
          || isBinaryOperation(tree, operator));

    return simpleExpression && condition.descendants().anyMatch(hasLiteralValue);
  }

}

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/IfConditionalAlwaysTrueOrFalseCheck.kt
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.checks

import org.jetbrains.kotlin.lexer.KtSingleValueToken
import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.KtBinaryExpression
import org.jetbrains.kotlin.psi.KtConstantExpression
import org.jetbrains.kotlin.psi.KtExpression
import org.jetbrains.kotlin.psi.KtIfExpression
import org.jetbrains.kotlin.psi.KtNameReferenceExpression
import org.jetbrains.kotlin.psi.KtUnaryExpression
import org.jetbrains.kotlin.psi.psiUtil.anyDescendantOfType
import org.sonar.check.Rule
import org.sonarsource.kotlin.api.AbstractCheck
import org.sonarsource.kotlin.plugin.KotlinFileContext

/**
 * Replacement for [org.sonarsource.slang.checks.IfConditionalAlwaysTrueOrFalseCheck]
 */
@Rule(key = "S1145")
class IfConditionalAlwaysTrueOrFalseCheck : AbstractCheck() {

    override fun visitIfExpression(expression: KtIfExpression, kotlinFileContext: KotlinFileContext) {
        if (isAlwaysTrueOrFalse(expression.condition!!)) {
            /** TODO try to use [org.jetbrains.kotlin.resolve.CompileTimeConstantUtils.canBeReducedToBooleanConstant] */
            kotlinFileContext.reportIssue(
                expression.condition!!,
                "Remove this useless \"if\" statement.")
        }
    }

    private fun isAlwaysTrueOrFalse(condition: KtExpression) =
        condition.isTrueValueLiteral() || condition.isFalseValueLiteral()
            || condition.isSimpleExpressionWithLiteral(KtTokens.ANDAND) { it.isFalseValueLiteral() }
            || condition.isSimpleExpressionWithLiteral(KtTokens.OROR) { it.isTrueValueLiteral() }

    /** Replacement for [org.sonarsource.slang.checks.utils.ExpressionUtils.isTrueValueLiteral] */
    private fun KtExpression.isTrueValueLiteral(): Boolean {
        val e = skipParentheses()
        return (e is KtConstantExpression && e.text == "true")
            || (e.isNegation() && (e as KtUnaryExpression).baseExpression!!.isFalseValueLiteral())
    }

    /** Replacement for [org.sonarsource.slang.checks.utils.ExpressionUtils.isFalseValueLiteral] */
    private fun KtExpression.isFalseValueLiteral(): Boolean {
        val e = skipParentheses()
        return (e is KtConstantExpression && e.text == "false")
            || (e.isNegation() && (e as KtUnaryExpression).baseExpression!!.isTrueValueLiteral())
    }

    /** Replacement for [org.sonarsource.slang.checks.utils.ExpressionUtils.isNegation] */
    private fun KtExpression.isNegation() =
        this is KtUnaryExpression && this.operationToken == KtTokens.EXCL

    /** Replacement for [org.sonarsource.slang.checks.IfConditionalAlwaysTrueOrFalseCheck.isSimpleExpressionWithLiteral] */
    private fun KtExpression.isSimpleExpressionWithLiteral(
        operation: KtSingleValueToken,
        hasLiteralValue: (KtExpression) -> Boolean,
    ) = isSimpleExpression(operation) && anyDescendantOfType(hasLiteralValue)

    private fun KtExpression.isSimpleExpression(operation: KtSingleValueToken): Boolean =
        when (val e = skipParentheses()) {
            is KtNameReferenceExpression -> true
            is KtConstantExpression -> true
            is KtBinaryExpression -> e.operationToken == operation &&
                e.left!!.isSimpleExpression(operation) && e.right!!.isSimpleExpression(operation)
            else -> false
        }
}
